Add enhancement to BitArray

t-imamichi opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [X] I have added the tests to cover my changes.
- [X] I have updated the documentation accordingly.
- [X] I have read the CONTRIBUTING document.
-->

### Summary

Add more utility methods to BitArray

- [x] transpose
- [x] concatenate along existing axis
- [x] concatenate along shots axis
- [x] concatenate along bits axis
- [x] slicing along existing axis
- [x] slicing along shots axis
- [x] slicing along bits axis (this helps marginalize counts/bitstrings)
- [x] compute expectation values of diagonal observables
- [x] reno

### Details and comments

t-imamichi committed:
Add transpose to BitArray

t-imamichi committed:
add concatenate

t-imamichi committed:
add marginalize and `__getitem__`

t-imamichi committed:
add expectation_value

t-imamichi committed:
add tests of concatenate

t-imamichi committed:
add tests for __getitem__

t-imamichi committed:
add tests for marginalize

t-imamichi committed:
add tests for expectation_value

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +351,133 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+
+        Raises:
+            ValueError: if ``axes`` don't match this bit array.
+            ValueError: if ``axes`` includes any indices that are out of bounds.
```
Minor convention consistency. Qiskit does some of both, but I think the V2 primitives lean a bit more toward the cap convention at this point.

```suggestion
            axes: Tuple or list of ints, optional. See
                `numpy.transpose
                <https://numpy.org/doc/stable/reference/generated/numpy.transpose.html>`_
                for the details.

        Returns:
            BitArray: A bit array with axes permuted.

        Raises:
            ValueError: If ``axes`` don't match this bit array.
            ValueError: If ``axes`` includes any indices that are out of bounds.
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +351,133 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+        Args:
+            axes: tuple or list of ints, optional. See
+                `numpy.transpose
+                <https://numpy.org/doc/stable/reference/generated/numpy.transpose.html>`_
```
Good idea pointing directly at NumPy. I'm wondering whether we should point at the ndarray method instead? The main difference is that the method form also accepts `*args`. Do you think that's something people would find helpful, or is it just a burden to have two different ways of specifying the transpose?

t-imamichi replied:
Thank you for pointing out `*args`. I forgot about it. I updated this method to accept integers in addition to a tuple of integers. As for the link, I like that of the ndarray method. I updated it too.

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +351,133 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+            if index < 0 or index >= self.num_bits:
+                raise ValueError(
+                    f"index {index} is out of bounds for the number of bits {self.num_bits}."
+                )
```
```suggestion
                )
        # This implementation introduces a temporary 8x memory overhead due to bit 
        # unpacking. This could be fixed using bitwise functions, at the expense of a
        # more complicated implementation.
```

t-imamichi committed:
make concatenate static method and add stack_shots

t-imamichi committed:
add stack_shots and stack_bits

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +368,200 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+
+    def marginalize(self, indices: int | Sequence[int]) -> "BitArray":
+        """Return a bit array marginalized over some indices of interests.
+
```
```suggestion
        """Return a bit array marginalized over some indices of interest.

        .. note::
            The convention used by this method is that the index ``0`` corresponds to the least-significant 
            bit in the :attr:`~array`, or equivalently the right-most bitstring entry as returned by 
            :meth:`~get_counts` or :meth:`~get_bitstrings`, etc.

            If this bit array was produced by a sampler, then an index ``i`` corresponds to the   
            :class:`~.ClassicalRegister` location ``creg[i]]`.
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +368,200 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+        return BitArray(self._array.transpose(axes), self.num_bits)
+
+    def marginalize(self, indices: int | Sequence[int]) -> "BitArray":
+        """Return a bit array marginalized over some indices of interests.
```
I'm second guessing whether ``marginalize`` is the right name for this method, because this class doesn't actually represent a distribution. I'm wondering if ``bit_slice`` might be better, or something along those lines. What do you or other reviewers think?

t-imamichi replied:
You are right. I like the name `bit_slice`.

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +368,200 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+            A bit array marginalized to bits of interests.
+
+        Raises:
+            ValueError: if there are any invalid indices to marginalize over.
```
```suggestion
            ValueError: If there are any invalid indices to marginalize over.
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +368,200 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+        arr, num_bits = _pack(arr)
+        return BitArray(arr, num_bits)
+
+    def expectation_value(
```
In the meeting we had discussed an array-valued version of this method. I'm not sure if it's possible to introduce such a thing in this PR because it would be blocked by making ObservableArray a public class. @chriseclectic may have advice.

t-imamichi replied:
Since `sample_expval_{float,complex}` returns a single expval, we need to extend it to allow multiple operators and return multiple expvals. It looks technically easy, but it might be good to be a separate PR.

As the first step, it might be nice to extend this `expectation_value` method to handle `ObservableArray` itself or a list generated by `ObservableArray.tolist()` without changing the rust code, i.e., run a loop in `expectation_value` method.

As the second step, we can update the rust code optimized for our input of `expectaion_value` method and remove the loop in `expectation_value` method.

https://github.com/Qiskit/qiskit/blob/5fb343f5f2faf5578dc0afcd6a3393fbaa0e0c17/crates/accelerate/src/sampled_exp_val.rs#L58-L70

ihincks replied:
One way to interpret the current implementation is that it's implementing the trivial broadcast rule. Ie, the "shape" of the input observable is `()`. We could look to extend this method by, in a future PR, accepting a non-trivial shape of observables, like an ObservablesArray or something coercible thereto.

With this in mind, what do you think about removing the `loc` argument, and instead returning the output of `sampled_expectation_value` for _every_ index of the shape. For example, if a bit array has shape (2,5), then the output of this method will be a float array of shape (2,5)? The extension to multiple observables would be to demand broadcastability between the observables array and the bit array.

Please let me know if you want an example.

t-imamichi replied:
Retuning an array of expval with the same shape makes sense to me. If users want to compute expvals of a subset of an array, they can slice off a part of their bit array using the new methods introduced in this PR.
Which idea do you have in mind, returning an array of expvals or a data bin of evs and stds as same as Estimator V2 output?

ihincks replied:
I would say just the array. The stds could be derived from the array in the usual way (`1-ev^2` divided by shots etc etc), if the user requires it. Perhaps this could be mentioned in the docstring?


```python
def expectation_values(self, operator: ObservablesLike) -> NDArray[np.float64]:
    """Compute the expectation value the operator for every entry of this bit array.

    Args:
         operator: The operator to take the expectation value of. 
    
    Returns:
        An array of expectation values whose shape matches this bit array's :attr:`~shape`.
    """
```

with the goal of moving to

```python
def expectation_values(self, operator: ObservablesArrayLike) -> NDArray[np.float64]:
    """Compute the expectation values of the provided observables, broadcasted against this bit array.

    Args:
         operator: The operator(s) to take the expectation value of. Must have a shape broadcastable with
            with this bit array.
    
    Returns:
        An array of expectation values whose shape is the broadcast shape of ``operator`` and this bit array.
        
    Raises:
        ValueError: If the provided operator does not have a shape broadcastable with this bit array.
    """
```

I wonder if we could get this second version into 1.1? Probably not enough time.

t-imamichi replied:
I implemented the observable array version in the latest commit. Could you take a look at it?

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -110,6 +128,9 @@ def __repr__(self):
         return f"BitArray({desc})"
 
+    def __getitem__(self, indices):
+        return BitArray(self._array[indices], self.num_bits)
```
This could run into issues if indices slices the last axis because it would change the number of bits.

```suggestion
        if isinstance(indices, tuple) and len(indices) >= self.ndim + 2:
            raise ValueError("BitArrays cannot be sliced along the bits axis, see marginalize() instead.")
        return BitArray(self._array[indices], self.num_bits)
```

ihincks reviewed (commented):
This looks very nice, thank you @t-imamichi

t-imamichi committed:
Apply suggestions from code review

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

t-imamichi committed:
fix lint

t-imamichi committed:
fix transpose to allow n ints

t-imamichi commented:
There are some topics to discuss as follows.

1. Names of methods and functions.
  - `concatenate`: join BitArray along BitArray's axes. I named it because the behavior is similar to `np.concatenate`.
  - `stack_shots`: join BitArray along shot axis.
  - `stack_bits`: join BitArray along bit axis.
  - `marginalize`: slice data along bit axis. I like Ian's suggestion `bit_slice` https://github.com/Qiskit/qiskit/pull/12158#discussion_r1571614531.
 
2. Location of the three functions to join BitArrays. I currently make them static methods of BitArray. So, users need to call `BitArray.concatenate`, for example. There is another option to make a module such as `qiskit.primitives.containers.utils`.

3. How to extend `expectation_value` method to deal with array of operators. https://github.com/Qiskit/qiskit/pull/12158#discussion_r1571619070

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+            ValueError: If ``axes`` don't match this bit array.
+            ValueError: If ``axes`` includes any indices that are out of bounds.
+        """
+        if len(axes) == 0:
+            axes = tuple(reversed(range(self.ndim)))
+        if len(axes) == 1 and isinstance(axes[0], Sequence):
+            axes = axes[0]
+        if len(axes) != self.ndim:
+            raise ValueError("axes don't match bit array")
+        for i in axes:
+            if i >= self.ndim or self.ndim + i < 0:
+                raise ValueError(
+                    f"axis {i} is out of bounds for bit array of dimension {self.ndim}."
+                )
```
There might be a couple of missing edge cases here that we should catch. For example, in the second conditional, axes[0] might not exist. Or, if the axes are not a permutation of range(self.ndim), ie something bis repeated. Maybe the intention was to rely on numpy.transpose for this last one?

t-imamichi replied:
As for the second conditional, I think `axes[0]` exists because `len(axes) == 1`.
If `axes[0]` does not exits, the first conditional catches it.
As for non-permutation case, I intended to rely on `np.transpose`.

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+        return BitArray(self._array.transpose(axes), self.num_bits)
+
+    def marginalize(self, indices: int | Sequence[int]) -> "BitArray":
+        """Return a bit array marginalized over some indices of interest.
```
This is unlikely to make a difference, but I think it's probably _slightly_ more clear that we'll be keeping the provided indices, rather than throwing them away.
```suggestion
        """Return a bit array marginalized to some indices of interest.
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+    def marginalize(self, indices: int | Sequence[int]) -> "BitArray":
+        """Return a bit array marginalized over some indices of interest.
+
+        .. note::
```
```suggestion

        This method is called ``marginalize`` because its contents typically represent
        randomly sampled bitstrings, in which case the output of this function represents
        samples from the marginal distribution. However, it could just as well have been 
        called ``slice_bits``.

        .. note::
```

t-imamichi replied:
Thank you for writing the note. I like your suggestion `bit_slice`. So, if there is no objection, I will rename this method name. Then, we can skip this note.

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+        Args:
+            bitarrays: The bit arrays must have (1) the same number of bits,
+                (2) the same number of shots, and
+                (3) the same shape, except in the dimension corresponding to axis
```
```suggestion
                (3) the same shape, except in the dimension corresponding to ``axis``
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+            axis: The axis along which the arrays will be joined. Default is 0.
+
+        Returns:
+            BitArray: The concatenated bit array.
```
```suggestion
            The concatenated bit array.
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+                    f"and the bit array at index {i} has {ba.ndim} dimension(s)."
+                )
+        if axis is None:
+            raise ValueError(f"axis {axis} must be a non-negative integer.")
```
Not obvious to me why None gets special treatment here. Perhaps left-over from a previous iteration? 

```suggestion
```

t-imamichi replied:
I added this treatment because `np.concatenate` allows `axis=None`, which flattens the array.
https://numpy.org/doc/stable/reference/generated/numpy.concatenate.html

`BitArray.concatenate` behaves almost same as `np.concatenate` except `axis=None`.
I didn't include the behavior because `BitArray.ravel` can do it already and I think it's nice to simplify `axis` by allowing only integers.
But I don't have a strong opinion. I'm fine to remove this treatment.

t-imamichi replied:
I made a mistake in the previous comment. I noticed that `BitArray` does not have `ravel` method.
If necessary, I will update `transpose` to accept None or add `ravel` method.

t-imamichi replied:
Ian told me a right way to ravel a bit array. So ignore my previous comments.

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+
+    @staticmethod
+    def stack_shots(bitarrays: Sequence[BitArray]) -> BitArray:
+        """Join a sequence of bit arrays along shots.
```
```suggestion
        """Join a sequence of bit arrays along the shots axis.
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+                and (2) the same shape.
+
+        Returns:
+            BitArray: The stacked bit array.
```
```suggestion
            The stacked bit array.
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+
+    @staticmethod
+    def stack_bits(bitarrays: Sequence[BitArray]) -> BitArray:
+        """Join a sequence of bit arrays along bits.
```
```suggestion
        """Join a sequence of bit arrays along the bits axis.
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+    @staticmethod
+    def stack_bits(bitarrays: Sequence[BitArray]) -> BitArray:
+        """Join a sequence of bit arrays along bits.
+
```
```suggestion

.. note::
    This method is equivalent to per-shot bitstring concatenation.

```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+
+        Args:
+            bitarrays: The bit arrays must have (1) the same number of shots,
+                and (2) the same shape.
```
```suggestion
            bitarrays: Bit arrays that have (1) the same number of shots,
                and (2) the same shape.
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+                and (2) the same shape.
+
+        Returns:
+            BitArray: The stacked bit array.
```
```suggestion
            The stacked bit array.
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+        return sampled_expectation_value(counts, operator)
+
+    @staticmethod
+    def concatenate(bitarrays: Sequence[BitArray], axis: int = 0) -> BitArray:
```
nit: to match CamelCase to snake_case because the class is `BitArray`. This happens a few times, but I'll just make the suggestion once.
```suggestion
    def concatenate(bit_arrays: Sequence[BitArray], axis: int = 0) -> BitArray:
```

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+                    f"and the bit array at index {i} has shape {ba.shape}."
+                )
+        data = np.concatenate([_unpack(ba) for ba in bitarrays], axis=-1)
+        data, num_bits = _pack(data)
```
Obvious to you I'm sure, but we can, in the future, work to eliminate the temporary 8x memory overhead here.

t-imamichi replied:
Yes, we need to think of the way to reduce the memory overhead. If it is hard to do it with numpy, rust would be an option.

ihincks reviewed (commented):
Thanks, these are all looking good. I've gone through it carefully and I don't expect to make any more significant comments.

t-imamichi commented:
@ihincks Thank you for reviewing the codes. I will update the codes following your suggestions.
Btw, I'm wondering what are the most intuitive names for various concatenating functions `concatenate`, `stack_bits`, `stack_shots` and slicing function `slice_bits` or `bit_slice`. Do you have suggestions?

ihincks commented:
> @ihincks Thank you for reviewing the codes. I will update the codes following your suggestions. Btw, I'm wondering what are the most intuitive names for various concatenating functions `concatenate`, `stack_bits`, `stack_shots` and slicing function `slice_bits` or `bit_slice`. Do you have suggestions?

For a consistent look-feel, I'd like to see:
```
stack_bits -> concatenate_bits
stack_shots -> concatenate_shots
marginalize -> slice_bits
```

This renaming suggests we are additionally missing (?)
```
slice_shots
```
This last one is is low priority, but would be nice if there's time.

t-imamichi committed:
Apply suggestions from code review

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

t-imamichi commented:
Thank you for your suggestions of the names. I have a question. There is `concatenate` function w/o suffix (along a bit array axis). Do you think we need to add any suffix?

I renamed method names following Ian's suggestions and updated `expectation_values` to return all expectation values (the first suggestion in https://github.com/Qiskit/qiskit/pull/12158#discussion_r1585330101). I'm thinking of how to extend it to `ObservableArray`.

t-imamichi committed:
Apply suggestions from code review

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

t-imamichi committed:
renamed expectation_value method with expectation_values to return all expectation values

t-imamichi committed:
update expectation_values to allow observable array

t-imamichi commented:
The previous commit c144db63 takes a single observable and  the latest commit a189ce4 allows an observable array.

t-imamichi commented:
I will think of slice_shots and write reno tomorrow.

t-imamichi committed:
fix lint

ihincks commented on qiskit/primitives/containers/bit_array.py:
```diff
@@ -347,3 +372,230 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+
+    def expectation_values(self, observables: ObservablesArrayLike) -> NDArray[np.float64]:
+        """Compute the expectation values of the provided observables, broadcasted against
+        this bit arrays.
```
```suggestion
        this bit array.
```

ihincks reviewed (commented):
The new expectation values looks good.

t-imamichi committed:
Update qiskit/primitives/containers/bit_array.py

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

t-imamichi committed:
updated expectation_values and added slice_shots

t-imamichi committed:
reorg tests and add some more cases

t-imamichi committed:
add reno and improved an error message

t-imamichi commented:
I added `slice_shots` and reno

t-imamichi committed:
Update qiskit/primitives/containers/bit_array.py

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

t-imamichi committed:
fix a test
